### ISSUE TITLE
Gradient checkpointing and Activated LoRA models

### ISSUE THREAD

**Issue by kgreenewald** [huggingface/peft#2826]
### System Info

NA

### Who can help?

I can potentially implement changes in a PR, but would like guidance on how maintainers would prefer this fixed. Previously, I worked with @githubnemo on building the Activated LoRA feature for PEFT.

### Reproduction

Activated LoRA in PEFT works by having forward hooks for each layer that allow an argument "alora_offsets" to be passed from the PeftModelforCausalLM.forward into the appropriate lora variant forward. 

I've discovered that with gradient checkpointing (at least in the simple DPO example below), the .backward operation does not save this alora_offsets input, does not re-run the forward hooks to recreate it, and hence uses alora_offsets = None in practice for the .backward computation. To be clear, no errors are thrown, just the computed gradients are zero. Also turning gradient_checkpointing=False fixes the issue (as you might expect).

There might be a lot of ways to try to fix this and keep support for gradient checkpointing. For instance, the simplest fix (but also pretty unsafe) would be to edit ALoraLinearVariant.forward in tuners/lora/variants.py by adding the following lines: 
```
alora_offsets = kwargs.get("alora_offsets", None) 
if alora_offsets is not None:
         module.alora_offsets = alora_offsets
 else:
         alora_offsets = module.alora_offsets
```
Here, the initial forward pass in the training loop will use the specified alora_offsets coming from the forward hooks and save it. In the gradient-checkpointed backward pass, alora_offsets will be None, but the original value would be recovered and used. While this does seem to work (I tried it), it feels unsafe to me.

Perhaps there could be a safer way of doing this, e.g. by hashing the layer input x and active adapter, and using that to store alora_offsets more safely, i.e. the saved alora_offsets will only be used if the hashed value matches a key in some offsets table. As a result, this should only trigger in a .backward scenario where the true alora_offsets got removed for some reason. 

What do you think (e.g. @githubnemo )? Happy to clarify anything and/or discuss further.

Example train script where you can see that the gradients are all 0 (fixed by turning off gradient checkpointing).
```
from peft import LoraConfig
import os
from datasets import load_dataset
from trl import DPOConfig, DPOTrainer
from transformers import AutoModelForCausalLM, AutoTokenizer


model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct",device_map="cuda", trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-0.5B-Instruct",padding_side='left', trust_remote_code=True)
model.pad_token_id = tokenizer.pad_token_id
train_dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train")
test_dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="test")
peft_config = LoraConfig(
            task_type="CAUSAL_LM",
                r=32,
                lora_alpha=32,
                lora_dropout=0.05,
                bias="none",
                alora_invocation_tokens = [151644], #this is from the chat template, i.e. the generation prompt
                target_modules=["q_proj", "k_proj", "v_proj","o_proj",
                                                    "gate_proj", "up_proj", "down_proj",
                                                        ],
)
training_args = DPOConfig(output_dir="Qwen2-0.5B-DPO", report_to="wandb",per_device_train_batch_size=4,
            per_device_eval_batch_size=4,
            learning_rate=5e-6,
            gradient_checkpointing=True, 
            gradient_accumulation_steps=2,
            eval_strategy="steps",   
            eval_steps=500,                
            logging_steps=50,
            max_length=2048,
            save_strategy="steps",
            save_steps=1000,
            save_total_limit=2,)
trainer = DPOTrainer(model=model, args=training_args, processing_class=tokenizer, train_dataset=train_dataset,peft_config=peft_config, eval_dataset=test_dataset)
trainer.train()
model.save_pretrained("./models/dpo/alora")
```

### Expected behavior

See above

**Comment by lastras** [comment]
Thank you @kgreenewald . Would appreciate feedback from the maintainers on how to approach this.

**Comment by githubnemo** [comment]
Hey @kgreenewald,

I've reviewed this and written a small reproducer without external trainers and datasets:

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import LoraConfig, get_peft_model
import copy

tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-3.2-1B")

inputs = tokenizer("this is a test", return_tensors='pt')
inputs['labels'] = inputs['input_ids'].clone()

base_model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-3.2-1B")
peft_model = get_peft_model(base_model, LoraConfig(
    task_type="CAUSAL_LM",
    alora_invocation_tokens=[inputs['input_ids'][0,0].item()]),
)

params = [(name, param) for name, param in peft_model.named_parameters() if param.requires_grad]

def grad_stat(params):
    grad_p = []
    nongrad_p = []
    for name, param in params:
        if param.grad.sum() > 0:
            grad_p.append((name, copy.deepcopy(param)))
        else:
            nongrad_p.append((name, copy.deepcopy(param)))
    return grad_p, nongrad_p

peft_model.train(True)

# without checkpointing
print('normal')

output = peft_model(**inputs)
output.loss.backward()

grad_params_normal, nongrad_params_normal = grad_stat(params)

for _, p in params:
    p.grad = None

# with checkpointing
print('with checkpointing')

peft_model.prepare_model_for_gradient_checkpointing(peft_model)
peft_model.gradient_checkpointing_enable()

output = peft_model(**inputs)
output.loss.backward()

grad_params_checkpointing, nongrad_params_checkpointing = grad_stat(params)

print(set([n for n, _ in grad_params_checkpointing]) == set([n for n, _ in grad_params_normal]))
```

Indeed the backward step has `alora_offsets == None` and returns different gradients. The problem is that we rely on `PeftModel.forward` as a starting point (where `alora_offsets` is injected). But with gradient checkpointing we have a second way in, so to say. And this pathway doesn't include the pre-forward hooks that inject the `alora_offset` value into `kwargs`.

So, essentially, PEFT forward hooks are generally broken with gradient checkpointing. The easy way out would be to say that we don't support gradient checkpointing + anything that relies on PEFT forward hooks. Depending on the needed implementation effort this is a valid option but I don't think we are there yet. We could also take the attribute approach you already thought about, but I agree, it seems like a bad idea to use the modules for global state. I'm not sure we can avoid having such a state completely, though, since `_enable_peft_forward_hooks` terminates way before the gradient checkpointing is done, so we need to store the alora offsets *somewhere* in the meantime. 

Maybe something like this in `LoraModel._enable_peft_forward_hooks` could work?
```python
        if alora_offsets is not None:
            for n, layer in self.named_modules():
                if isinstance(layer, GradientCheckpointingLayer) and layer.gradient_checkpointing:
                    # forward hook that does effectively what _enable_peft_forward_hooks does but at the time
                    # the gradient checkpointing layer is invoked, storing the kwargs in this hook handler instance.
                    # note that we still need to keep track of the hook handles but at least it is a bit more general.
                    def a(*args, **kwargs):
                        print("gradientcp forward pre_hook", args[0])
                        module = args[1]
                        if not hasattr(module, '_peft_forward_hooks'):
                            module._peft_forward_hooks = []
                        for submodule in module.modules():
                            if isinstance(submodule, LoraLayer):
                                module._peft_forward_hooks.append(
                                    submodule.register_forward_pre_hook(
                                        partial(_alora_offsets_pre_forward_hook, alora_offsets=kwargs['alora_offsets']),
                                        with_kwargs=True)
                                )
                    # backward hook to unregister any forward pre hook. quite similar to the statements below after the yield
                    def b(*args, **kwargs):
                        print("gradientcp backward hook", args[0])
                        module = args[1]
                        if hasattr(module, '_peft_forward_hooks'):
                            print("removing hooks")
                            for hook in module._peft_forward_hooks:
                                hook.remove()
                    layer.register_forward_pre_hook(partial(a, n, alora_offsets=alora_offsets))
                    layer.register_backward_hook(partial(b, n))
                if isinstance(layer, LoraLayer):
                    pre_forward = partial(_alora_offsets_pre_forward_hook, alora_offsets=alora_offsets)
                    handle = layer.register_forward_pre_hook(pre_forward, with_kwargs=True)
                    hook_handles.append(handle)
```

Let's iterate a bit on the pros and cons since I don't think this is a trivial change (but maybe I'm missing something as well).

**Comment by lastras** [comment]
@githubnemo Thank you. Am I understanding correctly that this has uncovered a more general issue with PEFT and gradient checkpointing, not confined to activated LoRA specifically?

**Comment by githubnemo** [comment]
> [@githubnemo](https://github.com/githubnemo) Thank you. Am I understanding correctly that this has uncovered a more general issue with PEFT and gradient checkpointing, not confined to activated LoRA specifically?

Yes, this is only relevant to activated LoRA insofar that it uses the forward hook mechanism to achieve passing the token offsets which seems incompatible with how gradient checkpointing works without further modification.

However, the impact of this mainly concerns aLoRA and mixed batch LoRA no other method currently employs this mechanism if I recall correctly.

**Comment by BenjaminBossan** [comment]
> and mixed batch LoRA

Just to clarify in case anyone comes across this: mixed batch LoRA is only for inference, so there is no issue there.

**Comment by lastras** [comment]
@githubnemo To clarify, we do see this as an issue that needs to be fixed; we are looking into possible solutions but as @kgreenewald said, we welcome your suggestions.

**Comment by githubnemo** [comment]
@lastras @kgreenewald with #2860 merged I think this issue is now resolved. If this is not the case, please feel free to re-open!

**Comment by lastras** [comment]
Ok will test. Thank you !!!!
